afr: make fsync post-op aware of inodelk count

itisravi opened the pull request:
Problem:
Since commit bd540db1e, eager-locking was enabled for fsync. But on
certain VM workloads wit sharding enabled, shard xlator keeps sending
fsync on the base shard. This can cause blocked inodelks from other
clients (including shd) to time out due to call bail.

Fix:
Make afr fsync aware of inodelk count and not delay post-op + unlock
when inodelk count > 1.

Note: We do not know yet why VMs go in to paused state because of the
blocked inodelks but this patch should be a first step in reducing the
occurence.

Updates: #2198
Change-Id: Ib91ebdd3101d590c326e69c829cf9335003e260b
Signed-off-by: Ravishankar N <ravishankar@redhat.com>

itisravi commented:
/run regression

pranithk commented:
Do you have statedumps of clients/bricks when this happens?

itisravi commented:
> Do you have statedumps of clients/bricks when this happens?

No, the stale locks were cleared with clear-locks by the support team. @xhernandez found fsyncs could lead to the inodelk calls call bailing. You can run the below test which he wrote, on the same file from two different clients on a sharded volume to hit the fsync blocking issue.

```c
#define _GNU_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <pthread.h>
#include <errno.h>

#define BLOCK_SIZE 512

static char buffer[BLOCK_SIZE] __attribute__((__aligned__(BLOCK_SIZE)));

static void *
worker(void *arg)
{
    char *name = arg;
    off_t off;
    unsigned int i;
    int fd;

    fd = open(name, O_WRONLY | O_CREAT | O_DIRECT, 0644);
    if (fd < 0) {
        printf("open() failed\n");
        exit(1);
    }

    i = 0;
    while (1) {
        off = /*((off_t)rand() << 9) +*/ (1ULL << 26);
        if (pwrite(fd, buffer, BLOCK_SIZE, off) != BLOCK_SIZE) {
            printf("write() failed %d\n", errno);
            exit(1);
	}
        if (fsync(fd) < 0) {
            printf("fsync() failed\n");
            exit(1);
        }
        printf("%u\n", i++);
        usleep(250000);
    }

    return NULL;
}

int main(int argc, char *argv[])
{
    pthread_t thread;
    int i;

    if (argc < 2) {
        return 1;
    }

    srand(time(NULL));

    for (i = 1; i < argc; i++) {
        if (pthread_create(&thread, NULL, worker, argv[i]) != 0) {
            printf("pthread_create() failed\n");
            exit(1);
        }
    }

    pthread_join(thread, NULL);

    return 0;
}
```

pranithk commented:
> > Do you have statedumps of clients/bricks when this happens?
> 
> No, the stale locks were cleared with clear-locks by the support team. @xhernandez found fsyncs could lead to the inodelk calls call bailing. You can run the below test which he wrote, on the same file from two different clients on a sharded volume to hit the fsync blocking issue.
> 
> ```
> #define _GNU_SOURCE
> 
> #include <stdio.h>
> #include <stdlib.h>
> #include <unistd.h>
> #include <sys/types.h>
> #include <sys/stat.h>
> #include <fcntl.h>
> #include <pthread.h>
> #include <errno.h>
> 
> #define BLOCK_SIZE 512
> 
> static char buffer[BLOCK_SIZE] __attribute__((__aligned__(BLOCK_SIZE)));
> 
> static void *
> worker(void *arg)
> {
>     char *name = arg;
>     off_t off;
>     unsigned int i;
>     int fd;
> 
>     fd = open(name, O_WRONLY | O_CREAT | O_DIRECT, 0644);
>     if (fd < 0) {
>         printf("open() failed\n");
>         exit(1);
>     }
> 
>     i = 0;
>     while (1) {
>         off = /*((off_t)rand() << 9) +*/ (1ULL << 26);
>         if (pwrite(fd, buffer, BLOCK_SIZE, off) != BLOCK_SIZE) {
>             printf("write() failed %d\n", errno);
>             exit(1);
> 	}
>         if (fsync(fd) < 0) {
>             printf("fsync() failed\n");
>             exit(1);
>         }
>         printf("%u\n", i++);
>         usleep(250000);
>     }
> 
>     return NULL;
> }
> 
> int main(int argc, char *argv[])
> {
>     pthread_t thread;
>     int i;
> 
>     if (argc < 2) {
>         return 1;
>     }
> 
>     srand(time(NULL));
> 
>     for (i = 1; i < argc; i++) {
>         if (pthread_create(&thread, NULL, worker, argv[i]) != 0) {
>             printf("pthread_create() failed\n");
>             exit(1);
>         }
>     }
> ```
> 
> ```
> pthread_join(thread, NULL);
> 
> return 0;
> ```
> 
> }

But VM workloads don't modify the file from same client. Shard is only supported for single writer usecases.

pranithk commented:
PS: Approach/Patch itself is fine. Curious to know why this happened/

itisravi commented:
> PS: Approach/Patch itself is fine. Curious to know why this happened/

Yet to figure it out. Either afr is missing sending unlock request or the sequence of blocked locks call bailing eventually is leading to a deadlock.

xhernandez commented:
The test program does more than necessary (I was testing more cases), but it's useful to reproduce the issue.

In a real workload, with a single writer, the other accesses that get blocked are caused by self-heal. I've seen several inodelk requests coming from SHD and glfsheal blocked for several hours.

This alone is not enough to explain how this causes problems to the VM (the VM should work normally, only self-heal should have issues). However I've seen that once the VM lock is released, a cascade of grants, disconnects and reconnects could happen (when SHD receives an answer for a request it's not already waiting, it closes the connection, which causes the just granted lock to be released by the cleanup procedure in brick side, which grants the following blocked inodelk, causing another disconnection, ...).

I think one possibility is that there's some race here that prevents one of the granted inodelk from being removed during disconnections, becoming a stale lock and blocking the VM from reacquiring it. However I haven't been able to reproduce it.

pranithk commented on xlators/cluster/afr/src/afr-inode-write.c:
```diff
@@ -2531,6 +2541,12 @@ afr_fsync(call_frame_t *frame, xlator_t *this, fd_t *fd, int32_t datasync,
     if (!local->xdata_req)
         goto out;
 
+    ret = dict_set_int32_sizen(local->xdata_req, GLUSTERFS_INODELK_COUNT, 0);
```
Any reason to go with this instead of GLUSTERFS_INODELK_DOM_COUNT with domain this->name like in inode-write?

itisravi replied:
I did it to release the locks irrespective of the domain (shd, data, metadata) the blocked locks are in, but thinking more, since `afr_fsync` only takes data domain lock,  we can limit the detection to that domain alone with `this->name`

pranithk commented on xlators/cluster/afr/src/afr-inode-write.c:
```diff
@@ -44,6 +44,9 @@ __afr_inode_write_finalize(call_frame_t *frame, xlator_t *this)
 
+    if ((local->op == GF_FOP_FSYNC) && (local->num_inodelks > 1))
+        local->transaction.disable_delayed_post_op = _gf_true;
+
```
You don't want the lock to be released in this case? Maybe we can make it similar to how writes handle conflicting locks. i.e. get inodelk-count for afr domain. If the number of locks is greater than 1 it will automatically take care of releasing and reacquiring locks.

itisravi replied:
> You don't want the lock to be released in this case?

In the function `afr_changelog_post_op` , if `afr_is_delayed_changelog_post_op_needed` returns false, locks will be released if lock->owner list is empty.

I think we could also use the current writes handling conflict logic but since `inode_ctx->afr_lock` was introduced only for writes, I did not want to modify it and used  `local->transaction.disable_delayed_post_op` which was added for fsync. Should we re-use it for fsync too?

pranithk replied:
Yes.

itisravi committed:
afr: make fsync post-op aware of inodelk count

Problem:
Since commit bd540db1e, eager-locking was enabled for fsync. But on
certain VM workloads wit sharding enabled, shard xlator keeps sending
fsync on the base shard. This can cause blocked inodelks from other
clients (including shd) to time out due to call bail.

Fix:
Make afr fsync aware of inodelk count and not delay post-op + unlock
when inodelk count > 1, just like writev.

Code is restructured so that any fd based AFR_DATA_TRANSACTION can be made
aware by setting GLUSTERFS_INODELK_DOM_COUNT in xdata request.

Note: We do not know yet why VMs go in to paused state because of the
blocked inodelks but this patch should be a first step in reducing the
occurence.

Updates: #2198
Change-Id: Ib91ebdd3101d590c326e69c829cf9335003e260b
Signed-off-by: Ravishankar N <ravishankar@redhat.com>

itisravi commented:
/run regression

itisravi commented:
Review requested please.
